packages/graphback-datasync/src/conflict/conflictEngine.ts

```diff
@@ -0,0 +1,48 @@
+import { GraphbackContext, metadataMap, FieldTransform } from '@graphback/core';
+import { ConflictStateMap } from "../util";
+
+const {fieldNames} = metadataMap;
+/**
+ * API for conflict detection and resolution engine
+ */
+export interface ConflictEngine {
+  /**
+   * checkForConflicts
+   * @param serverState document in DB
+   * @param clientState document received from client
+   * @returns ConflictStateMap if there is a conflict, undefined if not
+   */
+  checkForConflicts(serverState: any, clientState: any): ConflictStateMap | undefined 
+
+  /**
+   * checkForConflicts
+   * @param conflict ConflictStateMap received from checkForConflicts
+   * @returns resolved document or undefined if conflict cannot be resolved
+   */
+  resolveConflicts?(conflict: ConflictStateMap): any
+
+  /**
+   * updateState
+   * Supplies the updated At
+   * @param document document to be updated
+   * @returns document with updated state
+   */
+  updateState?(): FieldTransform[]
```

Comment:
This implementation was based on the implementations managing state (for example updating counter), but that is not reliable especially for providers like Mongo. We will not be able to `updateState` here as we actually do that outside using transformer for lastUpdated. 

If we going to do this we might actually need datasync to use it's own updatedAt implementation in updateState. This probably requires discussion but flagging it early

Reply:
Yes, I agree, my thoughts on this are pretty jumbled at the moment,
> We will not be able to updateState here as we actually do that outside using transformer for lastUpdated.

I was thinking about just that, what about using a function to transform document here:
https://github.com/aerogear/graphback/pull/1591/files/390b9b642048fddf3b586cc6ce24fd1480f47ce6#diff-be2392f5aaf3d58e02d7e9c6429f279cR151
I think it looks fair, will try it out now, but I have some concerns towards:
> If we going to do this we might actually need datasync to use it's own updatedAt implementation in updateState. 

So, it was kind of a mistake coupling that with datasync as it is vanilla provider stuff I guess, but anyhow datasync still needs updatedAt for delta queries, so what I'm thinking is that for the default strategy with timestamps, we might use a dummy updateState function which will not actually set anything, this way custom conflict engines can update state while not affecting `updatedAt`

What do you think about settling for this until we sort out #1590 ?